iib/workers/tasks/build_merge_index_image.py

```diff
@@ -0,0 +1,366 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+import logging
+import tempfile
+
+from iib.workers.api_utils import set_request_state
+from iib.workers.tasks.build import (
+    _add_ocp_label_to_index,
+    _build_image,
+    _cleanup,
+    _create_and_push_manifest_list,
+    _get_external_arch_pull_spec,
+    get_image_label,
+    _get_present_bundles,
+    _get_resolved_image,
+    _opm_index_add,
+    _prepare_request_for_build,
+    _push_image,
+    _update_index_image_build_state,
+    _update_index_image_pull_spec,
+    _verify_index_image,
+)
+from iib.workers.tasks.celery import app
+from iib.workers.tasks.utils import request_logger, run_cmd, set_registry_token
+
+
+__all__ = ['handle_merge_request']
+
+log = logging.getLogger(__name__)
+
+
+def _create_intermediate_index_image(
+    binary_image,
+    from_index,
+    ocp_version,
+    base_dir,
+    request_id,
+    arch,
+    overwrite_from_index_token=None,
+):
+    """
+    Create index image with the same data as old one, but with new binary image and ocp version.
+
+    :param str binary_image: binary image to be used by the new index image.
+    :param str from_index: index image, whose data will be contained in the new index image.
+    :param str ocp_version: ocp version which will be added as a label to the image.
+    :param str base_dir: base directory where operation files will be located.
+    :param int request_id: the ID of the IIB build request.
+    :param str arch: the architecture to build this image for.
+    :param str overwrite_from_index_token: the token used for overwriting the input
+        ``from_index`` image. This is required for non-privileged users to use
+        ``overwrite_from_index``. The format of the token must be in the format "user:password".
+    """
+    set_request_state(request_id, 'in_progress', 'Creating intermediate index image')
+    log.info('Creating intermediate index image with a new binary image')
+    _opm_index_add(base_dir, [], binary_image, from_index, overwrite_from_index_token)
+
+    _add_ocp_label_to_index(ocp_version, base_dir, 'index.Dockerfile')
+    _build_image(base_dir, 'index.Dockerfile', request_id, arch)
```

Comment:
Why not call build_image here for every arch present? Check out handle_add_request for reference.

Reply:
I thought that it would be more efficient to built only one arch for the intermediate stages. All archs are build for the final index image. Was that a bad thinking on my part?